api/registry.py

```diff
@@ -0,0 +1,342 @@
+    ConfigurationSetting,
+    ExternalIntegration,
+)
+from core.scripts import LibraryInputScript
+from core.util.http import HTTP
+from core.util.problem_detail import ProblemDetail
+
+from api.adobe_vendor_id import AuthdataUtility
+from api.config import Configuration
+from api.problem_details import *
+
+
+class RemoteRegistry(object):
+    """A circulation manager's view of a remote service that supports
+    the OPDS Directory Registration Protocol:
+
+    https://github.com/NYPL-Simplified/Simplified/wiki/OPDS-Directory-Registration-Protocol
+
+    In practical terms, this may be a library registry (which has
+    DISCOVERY_GOAL and wants to help patrons find their libraries) or
+    it may be a shared ODL collection (which has LICENSE_GOAL).
+    """
+    DEFAULT_LIBRARY_REGISTRY_URL = "https://libraryregistry.librarysimplified.org/"
+
+    def __init__(self, integration):
+        """Constructor."""
+        self.integration = integration
+
+    @classmethod
+    def for_integration_id(cls, _db, integration_id, goal):
+        """Find a LibraryRegistry object configured
+        by the given ExternalIntegration ID.
+
+        :param goal: The ExternalIntegration's .goal must be this goal.
+        """
+        integration = get_one(_db, ExternalIntegration,
+                              goal=goal,
+                              id=integration_id)
+        if not integration:
+            return None
+        return cls(integration)
+
+    @classmethod
+    def for_protocol_and_goal(cls, _db, protocol, goal):
+        """Find all LibraryRegistry objects with the given protocol and goal."""
+        for i in _db.query(ExternalIntegration).filter(
+            ExternalIntegration.goal==goal,
+            ExternalIntegration.protocol==protocol,
+        ):
+            yield cls(i)
+
+    @property
+    def registrations(self):
+        """Find all of this site's successful registrations with
+        this RemoteRegistry.
+
+        :yield: A sequence of Registration objects.
+        """
+        for x in self.integration.libraries:
+            yield Registration(self, x)
+
+
+class Registration(object):
+    """A library's registration for a particular registry.
+
+    The registration does not correspond to one specific data model
+    object -- it's a relationship between a Library and an
+    ExternalIntegration, and a set of ConfigurationSettings that
+    configure the relationship between the two.
+    """
+
+    # A library may be succesfully registered with a registry, or the
+    # registration may have failed.
+    LIBRARY_REGISTRATION_STATUS = u"library-registration-status"
+    SUCCESS_STATUS = u"success"
+    FAILURE_STATUS = u"failure"
+
+    # A library may be registered in a 'testing' stage or a
+    # 'production' stage. This represents the _library's_ opinion
+    # about whether it's ready for production. The library won't
+    # actually show up in production feeds until the _registry_ also
+    # thinks it should.
```

Comment:
How will the registry's stage be handled?

Reply:
Both the library and the registry stages will be stored on the registry side. Once we get it working I can have the registry send information about both stages as part of the library's informational OPDS feed, alongside more standard items such as the library's name. Then clients (whether they be SimplyE or the admin interface) will know the status of a given library.